python/ctranslate2/extensions.py

```diff
@@ -352,22 +479,24 @@ def _callback(step_result):
         }
     )
 
-    async_result = process_func(*args, **kwargs)[0]
+    async_results = process_func(*args, **kwargs)
 
     def _catch_exception():
         try:
-            async_result.result()
+            for result in async_results:
+                result.result()
         except Exception as e:
             step_results.put(e)
+        generation_done.set()
 
     thread = threading.Thread(target=_catch_exception, daemon=True)
     thread.start()
 
-    while True:
-        step_result = step_results.get()
-
-        if step_result is None:
-            break
+    while not step_results.empty() or not generation_done.is_set():
+        try:
+            step_result = step_results.get_nowait()
+        except queue.Empty:
+            continue
```

Comment:
I did not realise immediately that the code is now busy-waiting for new elements. Is there anyway to keep using `get` instead of `get_nowait`?

Reply:
I think you can still use the trick with `None`, and call `step_results.put(None)` where `generation_done.set()` is called. Would this work?

Reply:
Yeah it would work, I used continue instead of break, in the very unlikely case where the token would not have been put in the queue yet

Reply:
All callbacks are executed when the method `generate_batch` exits so this case will not happen.

Also using `continue` can introduce a race condition if `continue` and the loop condition are run before `generation_done.set()` is executed in the `_catch_exception` thread. In that case the program will indefinitely hangs in `step_results.get()`.

I think we should simply rollback to the previous construction:

```python
while True:
    step_result = step_results.get()

    if step_result is None:
        break
```

So the event `generation_done` may not be needed.

Reply:
okay, I came back to this version